### DISCUSSION TITLE
Missing Gateway and Service Events

### DISCUSSION THREAD

**Discussion by AugustMiller** [craftcms/commerce#4008] [Ideas]
### Description

Events are not currently emitted by `craft\commerce\base\Gateway` or `craft\commerce\services\Gateways` when saving a Gateway’s configuration.

Craft typically emits events from the service class when saving Project Config-bound changes (`beforeSave` when modifying values, then `afterSave` in a PC event handler). The _model_ events appear to be less consistently triggered…

### Steps to reproduce

1. Add a listener in a custom module:
    ```php
    use craft\events\ModelEvent;
    use craft\commerce\base\Gateway;
    use yii\base\Event;

    Event::on(
       Gateway::class,
       Gateway::EVENT_BEFORE_SAVE,
        function(craft\events\ModelEvent $event) {
            Craft::dd('Got it!');
        }
    );
    ```
2. Save a new or existing gateway
    

### Expected behavior

- Using the example code, the page is halted.
- I should also be able to listen to events from the service class, i.e. `craft\commerce\services\Gateways::EVENT_BEFORE_SAVE_GATEWAY`.

### Actual behavior

No event is triggered.

---

Originally reported [on Discord](https://discord.com/channels/456442477667418113/456442884258922529/1130921749068914788)!

**Comment by lukeholder** [comment]
Thanks for reporting. This is the same for some other project config based settings. Added this to the list for Commerce 5.x.

Others will include:

- [ ] Order Statuses
- [ ] Line Item Statuses
- [ ] Gateways
- [x] Emails
- [x] PDFs
- [x] Product Types

**Comment by lukeholder** [comment]
Keeping this issue open until all services get the improved event handlers. Tracking the services here: https://github.com/craftcms/commerce/issues/3224#issuecomment-1649328036

**Comment by linear** [comment]
<p><a href="https://linear.app/craftcms/issue/COM-273/add-more-gateway-and-services-events">COM-273 Add more Gateway and Services events</a></p>

**Comment by lukeholder** [comment]
https://github.com/craftcms/commerce/pull/3885 Created, but not sure of the utility of these events so they have been placed on the backlog at low priority. 

Happy to hear how others would want to use these events if they existed.

**Reply by billmn** [reply]
Hi @lukeholder,
it's been a while since the issue was reported on Discord, so I don't recall the details clearly. However, I do remember not being able to perform certain operations (e.g., HTTP calls before save) during the development of a custom Gateway.

These calls were necessary to integrate the payment provider requested by the client.

In my opinion, even just for consistency, it would be appropriate for the events to be triggered as indicated by @AugustMiller

**Reply by lukeholder** [reply]
@billmn Thanks, I have moved it up in priority. Thanks.
